nexus/src/app/instance.rs

```diff
@@ -264,6 +264,70 @@ pub(crate) struct InstanceEnsureRegisteredApiResources {
     pub(crate) authz_instance: nexus_auth::authz::Instance,
 }
 
+async fn normalize_ssh_keys(
+    datastore: &DataStore,
+    opctx: &OpContext,
+    ssh_public_keys: &Option<Vec<NameOrId>>,
+) -> Result<Option<Vec<NameOrId>>, Error> {
+    let actor = opctx
+        .authn
+        .actor_required()
+        .internal_context("loading current user's ssh keys for new Instance")?;
+    let (.., authz_user) = LookupPath::new(opctx, &datastore)
+        .silo_user_id(actor.actor_id())
+        .lookup_for(authz::Action::ListChildren)
+        .await?;
+
+    let ssh_keys = match ssh_public_keys {
+        Some(keys) => Some(
+            datastore
+                .ssh_keys_batch_lookup(opctx, &authz_user, keys)
+                .await?
+                .iter()
+                .map(|id| NameOrId::Id(*id))
+                .collect::<Vec<NameOrId>>(),
+        ),
+        None => None,
+    };
+    if let Some(ssh_keys) = &ssh_keys {
+        if ssh_keys.len() > MAX_SSH_KEYS_PER_INSTANCE.try_into().unwrap() {
+            return Err(Error::invalid_request(format!(
+                "cannot attach more than {} ssh keys to the instance",
+                MAX_SSH_KEYS_PER_INSTANCE
+            )));
+        }
+    }
+    Ok(ssh_keys)
+}
+
+async fn normalize_affinity_groups(
+    datastore: &DataStore,
+    opctx: &OpContext,
+    authz_project: &authz::Project,
+    groups: &Vec<NameOrId>,
+) -> Result<Vec<NameOrId>, Error> {
+    Ok(datastore
+        .affinity_groups_batch_lookup(opctx, authz_project, groups)
+        .await?
+        .iter()
+        .map(|id| NameOrId::Id(id.into_untyped_uuid()))
+        .collect::<Vec<NameOrId>>())
+}
+
+async fn normalize_anti_affinity_groups(
+    datastore: &DataStore,
+    opctx: &OpContext,
+    authz_project: &authz::Project,
+    groups: &Vec<NameOrId>,
+) -> Result<Vec<NameOrId>, Error> {
+    Ok(datastore
+        .anti_affinity_groups_batch_lookup(opctx, authz_project, groups)
+        .await?
+        .iter()
+        .map(|id| NameOrId::Id(id.into_untyped_uuid()))
+        .collect::<Vec<NameOrId>>())
+}
```

Comment:
Worth having these return just `Vec<Uuid>` since they're all IDs? I'm guessing that would require changes downstream that might be annoying, though it could also let you delete some annoying match statements.

Reply:
I would like to - this what my TODO comment is about below. Right now, `InstanceCreate` is a big struct of unvalidated user parameters, and we *do* perform validation on it, but don't transform the type before handing it to the saga machinery. I think we **should** transform the type (hence the TODO) but it needs to be a NameOrId for now.